### PR TITLE
JDK-8200337: Generalize see and link tags for user-defined anchors

### PR DESCRIPTION
**Pull request by hns**
Please review a a new feature to allow `@link`, `@linkplain` and `@see` tags to link to arbitrary URI fragments in the generated documentation (including in auxiliary `doc-files` documentation).

The changes in module `jdk.compiler` are mostly cleanup changes retained from earlier versions of the patch. The current proposed version uses a very simple change in `ReferenceParser` to avoid parsing the member name section of the reference when a non-member fragment is encountered.

The implementation introduces a new form of reference with a double hash mark (`##`) separator. This is a change from the previous implementation which also auto-recognized URI fragments and documentation paths by looking for `-` characters which are not allowed in member names. This feature was removed upon further consideration because it makes the feature more complex and less recognizable.
 
Links to auxiliary documentation files follow the same rules. They are recognized by looking for `/` characters in the fragment name. This means that ordinary `id` attribute values must not contain `/`, while auxiliary file paths must contain a `/` character. Both restrictions should be easy to sustain.

One thing that is difficult for this feature is to provide a good link label if no label is supplied in the tag. In contrast to program element names a fragment name does usually not make a good human readable name. The solution is to use the fragment name as default label text. I expect that the feature will usually be used with a user provided label.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8200337](https://bugs.openjdk.org/browse/JDK-8200337): Generalize see and link tags for user-defined anchors
 * [JDK-8294195](https://bugs.openjdk.org/browse/JDK-8294195): Generalize see and link tags for user-defined anchors (**CSR**)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10395/head:pull/10395` \
`$ git checkout pull/10395`

Update a local copy of the PR: \
`$ git checkout pull/10395` \
`$ git pull https://git.openjdk.org/jdk pull/10395/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10395`

View PR using the GUI difftool: \
`$ git pr show -t 10395`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10395.diff">https://git.openjdk.org/jdk/pull/10395.diff</a>

</details>

### CODE REVIEW

**Comment by jonathan-gibbons** [comment]
> One thing that is difficult for this feature is to provide a good link label if no label is supplied in the tag.

Generate an error if no label is provided.  Now is the time to set the rules.

**Comment by jonathan-gibbons** [comment]
>One thing that is difficult for this feature is to provide a good link label if no label is supplied in the tag. In contrast to program element names a fragment name does usually not make a good human readable name. The solution is to use the fragment name as default label text. I expect that the feature will usually be used with a user provided label.

There's cognitive dissonance here.

1. a fragment name does usually not make a good human readable name
2. the solution is to use the fragment name

At a minimum, we should give a warning, or even an error. We can use the `TagletWriterImpl.invalidTagOutput` to display something inline in the page. We should consider updating the spec to require that a label is given.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java [hunk 1]
```diff
@@ -410,11 +410,9 @@ private DCText inlineText(WhitespaceRetentionPolicy whitespacePolicy) throws Par
      * Matching pairs of {@literal < >} are skipped. The text is terminated by the first
      * unmatched }. It is an error if the beginning of the next tag is detected.
      */
-    // TODO: allowMember is currently ignored
-    // TODO: boolean allowMember should be enum FORBID, ALLOW, REQUIRE
     // TODO: improve quality of parse to forbid bad constructions.
     @SuppressWarnings("fallthrough")
-    protected DCReference reference(boolean allowMember) throws ParseException {
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java:417]: Why has this been removed?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java [hunk 2]
```diff
@@ -1237,7 +1231,7 @@ public DCTree parse(int pos) throws ParseException {
                 @Override
                 public DCTree parse(int pos) throws ParseException {
                     skipWhitespace();
-                    DCReference ref = reference(false);
+                    DCReference ref = reference();
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/DocCommentParser.java:1234]: You seem to be opening up the possibility that `@throws` can be followed by a member, which is a regression.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/ReferenceParser.java [hunk 3]
```diff
@@ -113,9 +111,12 @@ public Reference parse(String sig) throws ParseException {
 
         try {
             int slash = sig.indexOf("/");
+            int hash = sig.indexOf("#");
+            if (hash > -1 && slash > hash) {
+                // A slash following a hash is part of a doc-file path, not a module separator
+                slash = -1;
+            }
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/parser/ReferenceParser.java:118]: I thought we were backing off file references ...

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TagletWriterImpl.java [hunk 4]
```diff
@@ -481,24 +481,34 @@ private Content linkSeeReferenceOutput(Element holder,
...
-
-        if (refMemName == null && refMem != null) {
-            refMemName = refMem.toString();
+        String refFragment =   ch.getReferencedFragment(refSignature);
+
+        if (refFragment == null && refMem != null) {
+            refFragment = refMem.toString();
+        } else if (refFragment != null && refFragment.startsWith("#")) {
+            if (labelContent.isEmpty()) {
+                // A non-empty label is required for fragment and auxiliary file links
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TagletWriterImpl.java:490]: "auxiliary file links" ...

**Comment by jonathan-gibbons** [comment]
(repeating? after first message was lost...?)

I primarily reviewed the `DocCommentParser` level of the code.

Removing the `allowMember` parameter from the `reference` method seems like a retrograde step, since there are places where the method is called that require a specific kind of reference, such as requiring a type name after `@throws`.

If anything you should honor the `TODO` that you deleted there, and upgrade the `boolean allowMember` parameter into some sort of enum set that allows a caller to specify the kinds of reference that are acceptable in any given syntactic position.  After all, the new `##` syntax should only be permitted in `@see` and `{@link}` nodes.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/parser/ReferenceParser.java (hunk 3 above)
Comment (hns) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/parser/ReferenceParser.java]: Changed back to old code

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TagletWriterImpl.java (hunk 4 above)
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TagletWriterImpl.java]: Removed mention of auxiliary files.

File: test/langtools/jdk/javadoc/doclet/testSeeLinkAnchor/TestSeeLinkAnchor.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+        javadoc("-d", out.toString(),
+                "--module-source-path", src.toString(),
+                "--no-platform-links",
+                "--module", "m1,m2",
+                "m2/com.m2");
+
+        checkExit(Exit.OK);
+        checkOrder("m1/module-summary.html",
+            """
+                    <a href="../m2/com/m2/Class2.html#main-heading"><code>See main heading in Class2</code></a>""");
```
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSeeLinkAnchor/TestSeeLinkAnchor.java:114]: It seems like an anti-pattern to check the order of a single item!
Is this as intended?  Other examples, below, provide pairs.

**Review by jonathan-gibbons — APPROVED** [review]
Thanks for the new support in the doc-comment-parser for kinds of references. While docent may be able to make better checks, it can also be disabled, so it is good to catch things early where that is possible.

I noticed various additional new/improved doc comments as well: nice!

Check the following question in `TestSeeLinkAnchor.java`

File: test/langtools/jdk/javadoc/doclet/testSeeLinkAnchor/TestSeeLinkAnchor.java (hunk 5 above)
Comment (hns) [reply test/langtools/jdk/javadoc/doclet/testSeeLinkAnchor/TestSeeLinkAnchor.java:114]: Admittedly it was not a conscious choice to use `checkOrder` here, but I don't think it is an anti-pattern. Using `checkOrder` for a single string adds no overhead compared to `checkOutput`, and If we ever add additional string I think it is ok to do an ordered check by default.

Comment (hns) [reply test/langtools/jdk/javadoc/doclet/testSeeLinkAnchor/TestSeeLinkAnchor.java:114]: Looking at the `JavadocTester` source code, I notice that the remaining uses of `checkOutput` are numbered and marked as TODO, which supports the idea that `checkOrder` should be the default.

**Pull request closed without merging**